# Data Preprocessing: Visualisation

The first step to developing a machine learning model is to examine your data. Data examination and visualisation is important to highlight how features correlate, what type of variable each feature is (string, int, etc), and how many missing values are present. Depending on the dataset size and number of features, there are a few common strategies to help get an understanding of the set.

## Initial Printout

In a smaller dataset, a good approach is simply to print the set out:

In [9]:
import pandas as pd

df = pd.read_csv("titanic/train.csv")
df.to_string

<bound method DataFrame.to_string of      PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                              

In very large datasets, it can be impractical to load and display all the rows. You can instead call df.head() to print the first five rows, and df.info() to show the column type and number of non-null values per column.

In [6]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


From here, we can note a few things:
1) Some columns such as Cabin contain many missing values. If we were to indiscriminately filter out all rows with missing values, we would only be left with 204 rows, eliminating over 75% of the data.
2) Some columns contain unique values for every row. For example, every name in the dataset is unique. If we just plugged the Name column in as a feature to an ML model, the model wouldn't be able to draw any meaningful conclusions from that feature. This doesn't mean, however, that the column is useless. Information such as a person's title may correlate to whether or not they survived (we'll explore this later).
3) Some values will need to be converted to be useful. For example, the Sex column can easily be converted to be of type float, by assigning 'male' = 0.0, and 'female' = 1.0. This is useful information for when we begin cleaning the data.

## Visualisation
Here, we do quite a basic visualisation. While we haven't processed all the features yet, this heatmap can be a useful way to gain insight into how some features may interact with the survival rate (and with each other).

In [ ]:
df['Sex'] = df['Sex'].replace({'male': 0, 'female': 1})
df[['Sex']].head()